### ISSUE TITLE
use urllib3 instead of requests!

### ISSUE THREAD

**Issue by vincentsarago** [NASA-IMPACT/edl-credential-rotation#2]
Because we use `requests`, we need to build the lambda package (using docker). 

We could ease the deployment by using `urllib3` (which is provided in AWS Lambda runtime (as boto3)) and then use simple `aws_lambda.Code.from_inline`

🤷

**Comment by chuckwondo** [comment]
@vincentsarago, thanks for opening this issue.

It's very timely because I recently ran into _yet another_ copy of the [SessonWithRedirectHeaders](https://github.com/NASA-IMPACT/edl-credential-rotation/blob/6d287bc1793e05d831b246903c000a6947d1ac04/lambdas/handler.py#L9-L31) "in the wild."

This class has bugged me for years because it is _utterly insecure_, even though it is specifically intended to provide security, yet I've seen it replicated across a number of projects, and it is even duplicated in the `earthaccess` library.

**TL;DR:** the `SessonWithRedirectHeaders` is insecure because setting the Earthdata Login (EDL) creds (username/password) on the `auth` attribute of the session exposes the credentials _before_ the `rebuild_auth` method can strip them.

Perhaps you already knew this, but I've always wondered where the class originated. Yesterday, I finally found its origin, completely by accident. I might have found it sooner if I had bothered to ask someone. Utterly surprisingly, it comes directly from the [NASA Earthdata Login documentation for users](https://uat.urs.earthdata.nasa.gov/documentation/for_users/data_access/python)!

Unfortunately, as I mentioned, the code examples in the docs that use `requests` are _utterly insecure_. The first `requests` example is the code that is being blindly replicated all over the place because, at a quick glance, it _appears_ to be secure.

**Why is it insecure?** Since the EDL creds are set on the session's `auth` attribute, they are _blindly_ sent to the host of the _initial_ request, which is typically _not_ the EDL host (since the whole purpose of the full code is to request S3 credentials from some endpoint _outside_ of EDL). The `rebuild_auth` method is invoked _only_ upon _redirect_, which occurs _after_ the initial request, so it's too late at that point. **We've already sent the EDL creds to some non-EDL host.**

Of course, this is perhaps not a particularly egregious security hole since the non-EDL hosts are DAACs, so they are trustworthy hosts, but it's a security hole nonetheless.

Further, the `urllib` example on that same page is _utterly broken_ (even after fixing a couple of syntax and import errors). I had found this `urllib` code somewhere else a year or so ago (something that Doug Newman wrote up ages ago, now clearly copied from these docs), and had attempted to use it back then, but was unsuccessful.

However, having now discovered the origin of the `requests` and `urllib` code in the Earthdata Login docs, I decided to really get my hands dirty to discover _why_ the `urllib` code doesn't work, and _how_ to fix it, and I did!

I'll open a PR with a _working_ and _secure_ version that uses only `urllib`.

In a future enhancement, we should add retry logic. However, the current code that leverages `requests` does not perform retries anyway (`requests` supports retries, but does not perform retries by default), so an initial `urllib` version without retries would not be missing any functionality provided by the current code that uses `requests`.

cc: @sharkinsspatial, @ceholden

**Comment by sharkinsspatial** [comment]
@chuckwondo Is it possible to update the offending source EDL docs with your new urlliib example?

**Comment by chuckwondo** [comment]
> @chuckwondo Is it possible to update the offending source EDL docs with your new urlliib example?

Absolutely. I intend to open an Earthdata Login support ticket about all of this.
